Some set up
============
```{r setup_R, include=TRUE}
rm(list=ls())

library(knitr)
library(markdown)
library(shiny)

opts_chunk$set(error=FALSE, tidy=TRUE, warning=FALSE, 
               highlight=TRUE, cache=TRUE,comment=NA,
               dev=c('png','pdf'),fig.align='center', fig.show='asis',dpi=92)


options(stringsAsFactors = FALSE)
```
*****

Microarry pre-processing workflow for Illumina BeadArray data
=================================================================

## Author 
Dr Stephen Newhouse  
*Senior Bioinformatician*  
**NIHR Biomedical Research Centre for Mental Health**  
South London and Maudsley NHS Foundation Trust,  
Institute of Psychiatry,  
Kings College London,  
Box P092, 
De Crespigny Park,   
London SE5 8AF  
**email:-** stephen.newhouse@kcl.ac.uk  
**web:-** http://core.brc.iop.kcl.ac.uk/  

## Happy Microarry pre-processing!!!
**If you have some skill in R, all of this will be fairly straight forward**  
**Note that this is just one way of doing things**.  
This is all based on my experience, but I can guarantee that following the steps in this workflow will give you clean and robust data, and will also help identify and problems you may have with your data.  

*You will be surprised at some of the things your Omic's data will reveal....*

## R SCRIPTS AND TEMPLATE WORKFLOW
**R Workflow:-** http://git.brc.iop.kcl.ac.uk/snewhousebrc/sjnewhouse/blob/dev/GENE_EXPRESSION/Illumina_expression_workflow/illumina_gene_expression_workflow_preProcessing.Rmd  

**Custom R Functions:-** http://git.brc.iop.kcl.ac.uk/snewhousebrc/sjnewhouse/blob/dev/GENE_EXPRESSION/Illumina_expression_workflow/sjnewhouse_misc_R.R  

## GenomeStudio Gene Expression SOP
For those interested in how the raw data files were produced, and the standard workflow we follow.  

**GenomeStudio SOP:-** http://confluence.brc.iop.kcl.ac.uk:8090/display/PIP/Illumina+Gene+Expression+Chip+SOP+v1  

## How to get the worflow and scripts
This works...if it doesnt, then email me.

```{r wget_scripts, eval=FALSE}

# The Microarry pre-processing workflow for Illumina BeadArray data
wget http://git.brc.iop.kcl.ac.uk/snewhousebrc/sjnewhouse/blob/dev/GENE_EXPRESSION/Illumina_expression_workflow/illumina_gene_expression_workflow_preProcessing.Rmd  

# Associated functions and extra stuff
wget http://git.brc.iop.kcl.ac.uk/snewhousebrc/sjnewhouse/blob/dev/GENE_EXPRESSION/Illumina_expression_workflow/sjnewhouse_misc_R.R  

```
*****

Example workflow based on real data (GAP). 
------------------------------------------

Project Directory and Files
==============================
Data files and directories used in this workflow. This is here as a note to the reader/user.
```{r Files, eval=FALSE}

# Get workflow template, rename and set up for your project.
Copy illumina_gene_expression_workflow_preProcessing.Rmd to your project directory and rename as [MY_PROJECT_NAME].illumina_gene_expression_workflow_preProcessing.Rmd..or something that makes sense to you.

# Data Directory
"/media/D/expression/GAP_Expression"

# Data Files Exported from GenomeStudio
* control_probe_profile_Final_Report.txt
* Group_Probe_Profile_Final_Report.txt
* probe_annotation_Final_Report.txt
* sample_table_Final_Report.txt

# Required Input Files

# 1) Input for lumiR
* Sample_and_Control_Probe_Profile_FinalReport.txt 

# User Provided Data Files
# Use NA to record missing values in batch_info.txt
# Use Unknown to record missing values in pheno_info.txt 

# 2) Phenotype File
* pheno_info.txt
# REQUIRED COLUMNS (MATCH FORMAT SHOWN HERE ie SEX not Sex or Gender etc!):- "Sample.ID","SEX","GROUPS","TISSUE","PHENOTYPE","Study_ID"
# example 
# some proceeing of pheno_info** to make all UPPERCASE
head pheno_info.tmp;
cat pheno_info.tmp | sed '1q' > header;
cat pheno_info.tmp | sed '1,1d' | tr '[:lower:]' '[:upper:]'  > tmp;
cat header tmp > pheno_info_upper.txt;
mv pheno_info_upper.txt pheno_info.txt
head pheno_info.txt


# 3) Batch Information File
* batch_info.txt
# REQUIRED COLUMNS:- "Sample.ID","RIN","RNA_YIELD" and any other related batch info eg dates or processing. 
# example 
$ head batch_info.txt


# Naming Convensions
* All UPPERCASE
* SEX = MALE, FEMALE or UNKNOWN
* Missing data = NA for all data. The exceptioins are: SEX,GROUPS,PHENOTYPE, TISSUE. Use UNKNOWN

```

****

## load libs
```{r loadLibs, eval=TRUE, message=FALSE}
# Load libraries 
library(lumi)
library(annotate)
library(lumiHumanAll.db)
library(affy)
library(cluster)
library(impute)
library(WGCNA)
library(gplots)
library(limma)
library(vsn)
library(MBCB)
library(lumiHumanIDMapping)
library(scatterplot3d)
library(relaimpo)
library(plyr)
library(ggplot2)
library(flashClust)


# a little housekeeping for par() and subsequent plots
def.par <- par(no.readonly = TRUE)
par(def.par)
```

## load source file with some processing functions
email stephen.newhouse@kcl.ac.uk for code. This will all be on git soon.

```{r loadSource, eval=TRUE}

# path to gene expression processing scripts
path_to_scripts <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final"

# load 'em
source(paste(path_to_scripts,"/sjnewhouse_misc_R.R",sep=""))

ls()
```

****

## set project settings and I/O
User is asked to manually provide these options.  
This sets the working directoty, prjoect name, input and output files, along with qc options for transformation and normalisation methods.  
This project configuration data is written to *.csv file in your project directory.

```{r setProjectOptions, eval=TRUE}
# project directory
project_dir <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final"

# set working dir again
setwd(project_dir)

# project name
project_name <- "GAP_DL_FINAL"

# processing_date <- format(Sys.Date(),"%d_%m_%Y") 
processing_date <- format(Sys.time(), "%d_%m_%Y_%s")

# output directory for lumi process and plots
out_dir <- paste(project_dir,"/",project_name,"_lumi_processing_",processing_date ,sep="")

##out_dir <- paste(project_dir,"/","GAP_DL_FINAL_lumi_processing_27_01_2015_1422368240",sep="")



# make project pre-processing directory
make_dir_command <- paste(" if [ ! -e ",out_dir," ]; then mkdir ",out_dir,"; fi",sep="")

system( make_dir_command )

# genomestudio reports
gs_report  <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/Sample_and_Control_Probe_Profile_FinalReport.txt"
gs_probe   <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/Group_Probe_Profile_Final_Report.txt" 
gs_sample  <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/sample_table_Final_Report.txt" 
gs_control <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/control_probe_profile_Final_Report.txt" 
anno_table <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/probe_annotation_Final_Report.txt"

# sample information
# FILE NAME must contain : Sample.ID,SEX,GROUPS,TISSUE,PHENOTYPE,Study_ID
pheno_file <- "~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/pheno_info.txt" 

# batch information
tech_pheno_file <-"~/Documents/A_Year_1_PhD/GAP_Data/2015_01_27_GAP_preprocessing_pipeline_final/final_reports_genomestudio/batch_info.txt"

# detection call rate threshold
probe_det <- 80
sample_det <- 80

# flag for gender and sampleNetwork 
sex_check <- 1   # DO THIS!! I'm not providing an option to skip this
iac_check <- 1   # DO THIS!! I'm not providing an option to skip this
iac_sd_thrs <- 2 # 2 or 3

# Model based background correction method (MLE as default)
# All data should be background correceted. 
# The recomended methods is MBCB (Model-based Background Correction for Beadarray)
# URL http://www.bioconductor.org/packages/release/bioc/html/MBCB.html
mbcb_method <- "MLE"

# Transformation method
transform_method <- "log2" ## "vst" # log2, vst or both

# Normalisation method
norm_method <- "rsn" ## "rsn" # quantile, rsn, or both

# Folks that done stuff
analyst_email <- "stephen.newhouse@kcl.ac.uk"
analyst_name <-  "Stephen Newhouse"
lab_contact_email <- "charles.curtis@kcl.ac.uk"
lab_contact_name <- "Charle Curtis"
```

## write_project_settings_to_file

```{r write_project_settings_to_file, eval=TRUE}  
# write settings to file
project_settings <- data.frame(
  project_dir=project_dir,
  project_name=project_name,
  out_dir=out_dir,
  gs_report=gs_report,
  gs_probe=gs_probe,
  gs_sample=gs_sample,
  gs_control=gs_control,
  anno_table=anno_table,
  pheno_file=pheno_file,
  tech_pheno_file=tech_pheno_file,
  probe_det=probe_det,
  sample_det=sample_det,
  sex_check=sex_check,
  iac_check=iac_check,
  iac_sd_thrs=iac_sd_thrs,
  mbcb_method=mbcb_method,
  transform_method=transform_method,
  norm_method=norm_method,
  analyst_email=analyst_email,
  analyst_name=analyst_name,
  lab_contact_email=lab_contact_email,
  lab_contact_name=lab_contact_name
  )

# some data wrangling
project_settings <- as.data.frame(t(project_settings))
colnames(project_settings) <- "Project_Setting"
project_settings$Project_Variable <- rownames(project_settings)
project_settings <- project_settings[,c("Project_Variable","Project_Setting")]

# write table to out_dir
write.table(project_settings, file=paste(out_dir,"/",project_name,".project_settings.csv",sep=""),row.names=FALSE,quote=FALSE,sep=",")

# check settings
project_settings
```

****

BEGIN PRE-PROCESSING Raw Expression Data
=========================================

## 1. read raw gene expression data 
```{r lumiR, tidy=TRUE, eval=TRUE}

# raw input 
# This is the 1) Probe Profile, 2) Control Probe Profile and 3) Sample Table, Final Reports exported from GenomeStudio, all concatenated 
if(is.na(gs_report))  stop(" WARNING!: YOU HAVENT PROVIDED ANY DATA TO READ" )


# read raw gene expression data from genomestudio reports and create ExpressionSet

eset_raw <- lumiR( paste(gs_report), 
                  lib.mapping="lumiHumanIDMapping",
                  checkDupId=TRUE,
                  detectionTh = 0.01,
                  convertNuID=TRUE,
                  inputAnnotation=TRUE,
                  annotationColumn=c('PROBE_ID','CHROMOSOME','SYMBOL',
                                     'DEFINITION','ACCESSION','ENTREZ_GENE_ID',
                                     'PROBE_TYPE','PROBE_START','PROBE_SEQUENCE',
                                     'PROBE_CHR_ORIENTATION','PROBE_COORDINATES','CHROMOSOME',
                                     'TRANSCRIPT','ILMN_GENE','REFSEQ_ID',
                                     'UNIGENE_ID','SYMBOL','PROTEIN_PRODUCT'), QC=TRUE)

# check it
eset_raw

# n_expression_chips
n_expression_chips <- dim(eset_raw)[2]
cat("  WARNING!: The number of expression chips=[",dim(eset_raw)[2],"]","\r","\n")

# getChipInfo
chip_id <- getChipInfo(eset_raw)$chipVersion
chip_species <- getChipInfo(eset_raw)$species
chip_probes <- getChipInfo(eset_raw)$matchedProbeNumber
```

## 2. read in sample, pheno, batch and pData(eset) information

### read_gs_sample_info
This is the Sample Table from the Genomestudio Final Report tab.  
This contains lots of information realted to GS processing and intensity data. We use this data in GS to flag and remove "bad" chips/samples before exporting the final reports for processing in R.

```{r read_gs_sample_info, eval=TRUE}

# gs_sample
if(is.na(gs_sample))  stop(" WARNING!: YOU HAVENT PROVIDED ANY SAMPLE INFORMATION!!!" )

gs_sample

gs_sample_data <- read.table(paste(gs_sample) ,skip=8,as.is=T,fill=T,head=T,sep="\t")
rownames(gs_sample_data) <- gs_sample_data$Sample.ID
gs_sample_data <- gs_sample_data[,names(gs_sample_data)!="X"] 
gs_tech_var <- c("BIOTIN","CY3_HYB","HOUSEKEEPING","LABELING","LOW_STRINGENCY_HYB","NEGATIVE..background.","Noise")
sel_col <- colnames(gs_sample_data) %in% gs_tech_var
colnames(gs_sample_data) <- c( colnames(gs_sample_data[!sel_col]) ,paste("tech.",gs_tech_var,sep="") )
# added this as genomestudio likes to add mystery columns to the end of this report 
n_samples <- dim(gs_sample_data)[1]  # number of rows ie samples

# save it
save(gs_sample_data,file=paste(out_dir,"/",project_name,".eset_raw.gs_sample_data.RData",sep="") )
```

### read_pheno_info
This is basic phenotype information provided by the user.  
Required feilds are:- "Sample.ID","SEX","GROUPS","TISSUE","PHENOTYPE","Study_ID".  
The GROUPS field is used in the qc to determine sample outliers per GROUP and id probes that are detected in X percent per group.

```{r read_pheno_info, eval=TRUE}
#  pheno_file
if(is.na(pheno_file))  stop(" WARNING!: YOU HAVENT PROVIDED ANY PHENOTYPE INFORMATION!!!" )

pheno_file

pheno_dat <- read.table(paste(pheno_file), as.is=T,fill=T,head=T,sep="\t")
save(pheno_dat,file=paste(out_dir,"/",project_name,".eset_raw.pheno_dat.RData",sep="") )
has_pheno_cols <- c("Sample.ID","SEX","GROUPS","TISSUE","PHENOTYPE","Study_ID") %in% names(pheno_dat);
missing_pheno_cols <- "FALSE" %in% has_pheno_cols
if(missing_pheno_cols == "TRUE") stop(" WARNING!: YOU ARE MISSING ESSENTIAL SAMPLE INFORMATION! MAKE SURE YOUR PHENO_FILE HAS:- Sample.ID,SEX,GROUPS,TISSUE,PHENOTYPE,Study_ID !!!")
raw_n_pheno_dat <- dim(pheno_dat)[1] # number of rows ie samples

cat(" Running toupper() on PHENOTYPE, GROUP AND TISSUE variables to fix potential case issues","\r")
# fix case
pheno_dat$PHENOTYPE <- toupper(pheno_dat$PHENOTYPE)
pheno_dat$SEX <- toupper(pheno_dat$SEX)
pheno_dat$GROUPS <- toupper(pheno_dat$GROUPS)
pheno_dat$TISSUE <- toupper(pheno_dat$TISSUE)
# a quick looksee at counts
table(pheno_dat$PHENOTYPE)
table(pheno_dat$GROUPS)
table(pheno_dat$TISSUE)
table(pheno_dat$SEX)
```

### read_batch_info
```{r read_batch_info, eval=TRUE}
# tech_pheno_file
if(is.na(tech_pheno_file))  stop(" WARNING!: YOU HAVENT PROVIDED ANY BATCH INFORMATION!!!" )
tech_pheno_file
tech_pheno <- read.table(paste(tech_pheno_file),head=T,sep=" ")
tech_pheno$Sentrix.Barcode <- as.character(tech_pheno$Sentrix.Barcode)
str(tech_pheno)
rownames(tech_pheno) <- tech_pheno$Sample.ID
colnames(tech_pheno) <- paste("tech.",names(tech_pheno),sep="")
colnames(tech_pheno) <- c("Sample.ID", names(tech_pheno[,-1]))
save(tech_pheno,file=paste(out_dir,"/",project_name,".eset_raw.tech_pheno.RData",sep="") )
```

### read_pdata_info
This is the actual list of chips present in the expression set object.  
This is the base starting point for all subsequent merges of sample information, and determines the chip or sample order.

```{r read_pdata_info, eval=TRUE}

# get pData()
eset_samples <- pData(eset_raw)

# add chip order and flad for "has expression data" to eset_samples (pData)
eset_samples$has_expression <- 1

# making chip_oder columm
eset_samples$chip_order <- 1:dim(eset_samples)[1]
save(eset_samples,file=paste(out_dir,"/",project_name,".eset_raw.pData_samples.RData",sep="") )
```

### quick_compare_pdata_pheno_batch_gs_sample_info
```{r quick_compare_pdata_pheno_batch_gs_sample_info, eval=TRUE}
# col names
names(eset_samples) 
names(gs_sample_data)
names(pheno_dat)
names(tech_pheno)

# head
head(eset_samples) 
head(gs_sample_data)
head(pheno_dat)
head(tech_pheno)

# quick check
# these should all have the same number of rows or samples!
dim(eset_samples);dim(gs_sample_data);dim(pheno_dat);dim(tech_pheno)

# Venn of Sample.ID
ex <- eset_samples$sampleID
pp <- pheno_dat$Sample.ID
tt <- tech_pheno$Sample.ID
venninput  <-list(ArrayExpression=ex,Batch_Info=tt,Pheno_Info=pp)
venn(venninput)
#dev.off()
```

## 2 check for duplicate Study_ID
```{r check_for_ duplicate Study_ID, eval=TRUE}
# check for duplicate Study_ID
tab_id <- table(pheno_dat$Study_ID)
tab_id_df <- as.data.frame(tab_id)
colnames(tab_id_df) <- c("Study_ID","Freq")
dupe_samples <- subset(tab_id_df,tab_id_df$Freq>=2) 
if(dim(dupe_samples)[1]>1) {
  cat("  WARNING!: You have duplicate Study_IDs. N=[",dim(dupe_samples)[1],"]","\r","\n") 
  }
# show dupes
dupe_samples
# save to file
write.table(dupe_samples,file=paste(out_dir,"/",project_name,".dupe_Study_IDs.txt",sep=""),sep="\t",quote=FALSE,row.names=FALSE)
# n_unique_study_id
n_unique_study_id <- length(unique(tab_id_df$Study_ID))
cat("  WARNING!: The number of unique Study_Ids=[",n_unique_study_id,"]","\r","\n")

```


## 3. check eset_samples, sample & batch info Sample.ID's match in names and numbers & merge all
```{r Merge_Sample_Pheno_Batch_Info, eval=TRUE}
cat(" Megreing pdata, pheno data, batch information adn genomestudio samople data","\r","\n")
# 1. merge eset_samples with pheno_dat. 
# Keep ALL overlaps only
eset_pheno_merge <- merge(eset_samples,pheno_dat,by.x="sampleID",by.y="Sample.ID" )
eset_pheno_merge <- eset_pheno_merge[order(eset_pheno_merge$chip_order),]
dim(eset_samples);dim(eset_pheno_merge) # check size

# 2. merge eset_pheno_merge with tech_pheno
eset_pheno_batch_merge <- merge(eset_pheno_merge,tech_pheno,by.x="sampleID",by.y="Sample.ID" )
eset_pheno_batch_merge <- eset_pheno_batch_merge[order(eset_pheno_batch_merge$chip_order),]
dim(eset_samples);dim(eset_pheno_merge);dim(eset_pheno_batch_merge) # check size

# 3. merge all with genomestudio final report
eset_pheno_batch_gs_merge <- merge(eset_pheno_batch_merge,gs_sample_data,by.x="sampleID",by.y="Sample.ID" )
eset_pheno_batch_gs_merge <- eset_pheno_batch_gs_merge[order(eset_pheno_batch_gs_merge$chip_order),]

# final look at numbers in each merged data set
dim(eset_samples);dim(eset_pheno_merge);dim(eset_pheno_batch_merge);dim(eset_pheno_batch_gs_merge)

# names
names(eset_pheno_batch_gs_merge)

# looksee
head(eset_pheno_batch_gs_merge)

# quick visual check to make sure chip order is intact
plot(eset_pheno_batch_gs_merge$chip_order,pch=20, cex=0.6, main="this should be a straight line")

```

## 4. Subset raw ExpressionSet to matched/complete Sample.IDs & Update pData() slot.
Here we subset the expression data to those chips/samples that have phenotype and batch data.  

```{r subsetEsetRaw, eval=TRUE}

# samples in gene expression data
samples_eset <- pData(eset_raw)$sampleID
length(samples_eset)

# samples with complete data
samples_complete_data <- eset_pheno_batch_gs_merge$sampleID
length(samples_complete_data)  

# samples to remove
samples_to_remove <- (samples_eset %in% samples_complete_data)==FALSE  
samples_to_remove <- pData(eset_raw)$sampleID[samples_to_remove]
length(samples_to_remove)  


# rename eset_raw & save
eset_raw_preqc <- eset_raw
cat(" saving eset_raw before any qc takes place - this will be the pure un altered raw data file, subseted to samples with pheno data","\r","\n")
cat(" File=[",paste(out_dir,"/",project_name,".eset_raw_preqc.RData",sep=""),"]","\r","\n")
save(eset_raw_preqc, file=paste(out_dir,"/",project_name,".eset_raw_preqc.RData",sep=""))


# subset eset_raw
eset_raw <- removeSamples_eset_lumi(eset=eset_raw_preqc,sampleRemove=samples_to_remove)
eset_raw


# update pData
old_pdata <- pData(eset_raw)
old_pdata$old_order <- 1:dim(old_pdata)[1]

# merge with eset_pheno_batch_gs_merge
new_pdata <- merge(old_pdata, eset_pheno_batch_gs_merge, by.x="sampleID", by.y="sampleID", all=TRUE,sort=FALSE)
new_pdata <- new_pdata[order(new_pdata$old_order),]

# remove columns old_order has_expression chip_order
new_pdata <- new_pdata[,-c(2,3,4)]

# update rownames
rownames(new_pdata) <- new_pdata$sampleID

# update pData slot
pData(eset_raw) <- new_pdata
dim(pData(eset_raw))

# check it
eset_raw 

## n_expression_chips_with_data
n_expression_chips_with_data <- dim(eset_raw)[2]

cat(" Number of chips with complete Phenotype and Batch data=[",n_expression_chips_with_data,"]","\r","\n")
    
```

## 5. Add nuID to fData
nuID is a stable probe id and should be used over and above your standard gene ids - illumina has a habit of changing probe names and sequences!.

```{r updateFeatureData, eval=TRUE}
 #Add nuID to fData
cat(" Add nuID to fData","\r","\n")
fData(eset_raw)$nuID <- rownames(fData(eset_raw))
head(fData(eset_raw))
```

## 6. Save updated raw ExpressionSet eset_raw
```{r saveEsetRaw, eval=TRUE}
# Save updated raw ExpressionSet eset_raw
save(eset_raw, file=paste(out_dir,"/",project_name,".eset_raw.RData",sep=""))
```

## 7. Write data files to out_dir for eset_raw
```{r writeEsetRawData, eval=TRUE}
# Write data files to out_dir for eset_raw
write_expression_files(eset=eset_raw,outfile=paste(out_dir,"/",project_name,".eset_raw",sep="") )
```

QC plots on `eset_raw`
-----------------------------------------------------------
### basic_qc_plot_lumi
```{r basic_qc_plot_lumi_eset_raw, fig.height=7, fig.width=10,eval=TRUE}
# basic plots
# plot to screen
basic_qc_plot_lumi(eset_raw)
par(def.par)
```

```{r basic_qc_plot_lumi_eset_raw_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_raw.basic_qc_plot_lumi.pdf",sep=""),width=11,height=8)
basic_qc_plot_lumi(eset_raw)
dev.off()
par(def.par)
```

### coloured_dendrogram_lumi
```{r coloured_dendrogram_lumi_eset_raw, fig.height=7, fig.width=10,eval=TRUE}
# coloured_dendrogram_lumi
# plot to screen
par(mar=c(5,20,5,5))
coloured_dendrogram_lumi(eset_raw)
par(def.par)
```

```{r coloured_dendrogram_lumi_eset_raw_pdf, eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_raw.coloured_dendrogram_lumi.pdf",sep=""),width=11,height=8)
coloured_dendrogram_lumi(eset_raw)
dev.off()
par(def.par)

```

### pca_plot_lumi
```{r pca_plot_lumi_eset_raw, fig.height=7, fig.width=7,eval=TRUE}
# PCA plots
# plot to screen
pca_plot_lumi(eset_raw)
par(def.par)
```

```{r pca_plot_lumi_eset_raw_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_raw.pca_plot_lumi.pdf",sep=""),width=7,height=7)
pca_plot_lumi(eset_raw)
dev.off()
par(def.par)

```

### SampleNetwork Plots
```{r sampleNetwork_plot_all_lumi_eset_raw, fig.height=10, fig.width=10,eval=TRUE}
# SampleNetwork Plots
# plot to screen
sampleNetwork_plot_all_lumi(eset_raw, colBy="chip");
sampleNetwork_plot_all_lumi(eset_raw, colBy="group");
par(def.par)
```


```{r sampleNetwork_plot_all_lumi_eset_raw_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_raw.sampleNetwork_plot_all_lumi.pdf",sep=""),width=8,height=8)
sampleNetwork_plot_all_lumi(eset_raw, colBy="chip");
sampleNetwork_plot_all_lumi(eset_raw, colBy="group");
dev.off()
par(def.par)

```

Basic fundamentalNetworkConcepts `eset_raw`
-------------------------------------------------------------------

## SampleNetwork on eset_raw for all samples as a first pass
```{r SampleNetwork_raw_data_all, eval=TRUE}
datExprs <- exprs(eset_raw)
samle_names <- sampleNames(eset_raw)
IAC=cor(datExprs, method="p",use="p")
diag(IAC)=0
A.IAC=((1+IAC)/2)^2  ## ADJACENCY MATRIX
# fundamentalNetworkConcepts
FNC=fundamentalNetworkConcepts(A.IAC) ## WGCNA
K2=FNC$ScaledConnectivity
Z.K=round( (K2-mean(K2))/sd(K2) ,3)
Z.C=round( (FNC$ClusterCoef-mean(FNC$ClusterCoef))/sd(FNC$ClusterCoef),3)
# cor K,C
rho <- signif(cor.test(Z.K,Z.C,method="s")$estimate,2)
rho_pvalue <- signif(cor.test(Z.K,Z.C,method="s")$p.value,2)
# Z.K_outliers 
Z.K_outliers <- Z.K < -iac_sd_thrs
Z.K_outliers <- names(Z.K_outliers[Z.K_outliers==TRUE])
n_outliers <- length(Z.K_outliers)
mean_IAC <- signif(mean(IAC[upper.tri(IAC)]),2)
min_Z.K <- min(Z.K)
cat(" Number of Z.K outliers=[", n_outliers,"]","\r","\n")
cat(" mean_IAC=[", mean_IAC,"]","\r","\n")
cat(" cor(Z.k,Z.C)=[", rho,"] P=[",rho_pvalue,"]","\r","\n")
# print chip ids
Z.K_outliers
# get these bad samples from pData and update pdata to include these metrics
pData(eset_raw)$Z.K_eset_raw <- Z.K
pData(eset_raw)$Z.C_eset_raw <- Z.C
pData(eset_raw)$cor_Z.K.Z.C_eset_raw <- rho
pData(eset_raw)$cor_p_Z.K.Z.C_eset_ra <- rho_pvalue
# take a look at these outliers
samples_out <- pData(eset_raw[,Z.K_outliers])
head(samples_out)
# wrtite csv
write.table(samples_out, file=paste(out_dir,"/",project_name,".eset_raw_Z.K_outliers.csv",sep=""),sep=",",quote=FALSE, row.names=FALSE)
# looksee
table(samples_out$GROUPS)
table(samples_out$PHENOTYPE)
table(samples_out$SEX)
table(samples_out$tech.Sentrix.Barcode)
# get and save list of Z.K outliers
eset_raw_Z.K_outliers <- Z.K_outliers
save(eset_raw_Z.K_outliers  , file=paste(out_dir,"/",project_name,".eset_raw_Z.K_outliers.RData",sep=""))
```


Save `eset_raw`
-------------------------------------------------------------------

```{r save_eset_raw_after_basic_sampleNetwork, eval=TRUE}
# Save updated raw ExpressionSet eset_raw
save(eset_raw, file=paste(out_dir,"/",project_name,".eset_raw.RData",sep=""))
```

```{r write_eset_raw_after_basic_sampleNetwork, eval=TRUE}
# Write data files to out_dir for eset_raw
write_expression_files(eset=eset_raw,outfile=paste(out_dir,"/",project_name,".eset_raw",sep="") )
```


****

MBCB (Model-based Background Correction for Beadarray)
=======================================================
Background correction! Do not skip this step.
If you have data from Genomestudio you will have NEGATIVE BEAD expression levels to use for background correction.  
http://www.ncbi.nlm.nih.gov/pubmed/18450815.  
Nucleic Acids Res. 2008 Jun;36(10):e58. doi: 10.1093/nar/gkn234. Epub 2008 May 1.  
**Enhanced identification and biological validation of differential gene expression via Illumina whole-genome expression arrays through the use of the model-based background correction methodology**.Ding LH, Xie Y, Park S, Xiao G, Story MD.  
The alternative is Robust multi-array (RMA) if you dont have NEGATIVE BEAD expression levels. Noe example code is provided for this.

## 1. Run mbcb.correct(signal,negCon,npBool=FALSE,mleBool=TRUE,isRawBead=FALSE)
```{r runMBCB, eval=TRUE}
# Run Model-based Background Correction for Beadarray : method = MLE
# mbcb.correct(signal,negCon,npBool=FALSE,mleBool=TRUE, isRawBead=FALSE)
eset_bg <- bgcor_mbcb(eset=eset_raw, outfile=paste(out_dir,"/",project_name,".eset_bg",sep="") )
eset_bg 
```

## 2. SAVE BACKGROUND CORRECTED DATA 
```{r save_eset_bg, eval=TRUE}
save(eset_bg, file=paste(out_dir,"/",project_name,".eset_bg.RData",sep="")  , compress=T)
```

## 3. Write data files to out_dir for eset_bg
```{r write_eset_bg_Data, eval=TRUE}
# Write data files to out_dir for eset_raw
write_expression_files(eset=eset_bg,outfile=paste(out_dir,"/",project_name,".eset_bg",sep="") )
```

QC Plots of `eset_bg`
---------------------------------------------------------

## basic_qc_plot_lumi 
```{r basic_qc_plot_lumi_eset_bg, fig.height=7, fig.width=10,eval=TRUE}
# basic plots
# plot to screen
basic_qc_plot_lumi(eset_bg)
par(def.par)
```

```{r basic_qc_plot_lumi_eset_bg_pdf, eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_bg.basic_qc_plot_lumi.pdf",sep=""),width=11,height=8)
basic_qc_plot_lumi(eset_bg)
dev.off()
par(def.par)

```

## coloured_dendrogram_lumi 
```{r coloured_dendrogram_lumi_eset_bg, fig.height=7, fig.width=10,eval=TRUE}
# coloured_dendrogram_lumi
# plot to screen
par(mar=c(5,20,5,5))
coloured_dendrogram_lumi(eset_bg)
par(def.par)
```

```{r coloured_dendrogram_lumi_eset_bg_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_bg.coloured_dendrogram_lumi.pdf",sep=""),width=11,height=8)
coloured_dendrogram_lumi(eset_bg)
dev.off()
par(def.par)

```

## pca_plot_lumi 
```{r pca_plot_lumi_eset_bg, fig.height=7, fig.width=7,eval=TRUE}
# PCA plots
# plot to screen
pca_plot_lumi(eset_bg)
```

```{r pca_plot_lumi_eset_bg_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_bg.pca_plot_lumi.pdf",sep=""),width=7,height=7)
pca_plot_lumi(eset_bg)
dev.off()
par(def.par)

```

## SampleNetwork Plots 
```{r sampleNetwork_plot_all_lumi_eset_bg, fig.height=10, fig.width=10,eval=TRUE}
# SampleNetwork Plots
# plot to screen
sampleNetwork_plot_all_lumi(eset_bg, colBy="chip");
sampleNetwork_plot_all_lumi(eset_bg, colBy="group");
par(def.par)
```

```{r sampleNetwork_plot_all_lumi_eset_bg_pdf,eval=TRUE}
pdf(file=paste(out_dir,"/",project_name,".eset_bg.sampleNetwork_plot_all_lumi.pdf",sep=""),width=8,height=8)
sampleNetwork_plot_all_lumi(eset_bg, colBy="chip");
sampleNetwork_plot_all_lumi(eset_bg, colBy="group");
dev.off()
par(def.par)

```